### ISSUE TITLE
Non-hexadecimal digit found

### ISSUE THREAD

**Issue by justindowning** [nathancahill/mimicdb#2]
@nathancahill First off, thanks for making this!

I'm trying to understand the best workflow for incorporating mimicdb in my tooling. Basically, I'm trying to syncronize two S3 buckets in the same region. To do this, I'm doing the following:

1. Get a connection to S3
2. Sync the connection with mimicdb/redis (`conn.sync(source_bucket_name, dest_bucket_name)`)
3. Iterate over keys in source_bucket and `copy_key` those that are not present in the dest_bucket

The problem (I think) I'm having is during the sync.
```
$ ./usemimic.py 
Traceback (most recent call last):
  File "./usemimic.py", line 78, in <module>
    copy_bucket(aws_key, aws_secret_key, args)
  File "./usemimic.py", line 56, in copy_bucket
    src_bucket.sync()
  File "/usr/local/lib/python2.7/dist-packages/mimicdb/s3/bucket.py", line 161, in sync
    for key in self.list(force=True):
  File "/usr/local/lib/python2.7/dist-packages/mimicdb/s3/bucket.py", line 112, in list
    for key in super(Bucket, self).list(*args, **kwargs):
  File "/usr/local/lib/python2.7/dist-packages/boto/s3/bucketlistresultset.py", line 34, in bucket_lister
    encoding_type=encoding_type)
  File "/usr/local/lib/python2.7/dist-packages/mimicdb/s3/bucket.py", line 72, in get_all_keys
    return super(Bucket, self).get_all_keys(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/boto/s3/bucket.py", line 472, in get_all_keys
    '', headers, **params)
  File "/usr/local/lib/python2.7/dist-packages/mimicdb/s3/bucket.py", line 142, in _get_all
    key.name = key.name
  File "/usr/local/lib/python2.7/dist-packages/mimicdb/s3/key.py", line 60, in name
    self._load_meta(meta['size'], meta['md5'])
  File "/usr/local/lib/python2.7/dist-packages/mimicdb/s3/key.py", line 38, in _load_meta
    self.md5 = md5
  File "/usr/local/lib/python2.7/dist-packages/boto/s3/key.py", line 173, in _set_md5
    self.local_hashes['md5'] = binascii.a2b_hex(value)
TypeError: Non-hexadecimal digit found
```

Any thoughts here or perhaps an idea how I can catch this instead of stopping the script? Thanks!

**Comment by nathancahill** [comment]
Interesting, looking into this. Will report back with some tests. Thanks!

**Comment by justindowning** [comment]
@nathancahill have you had some time to look into this?

**Comment by jakubincloud** [comment]
MimicDB does not support keys uploaded as multi part. In this case Amazon doesn't store MD5 in ETag parameter.

The problem is that you use ETag to set md5
bucket.py:140 
mimicdb.backend.hmset(tpl.key % (self.name, key.name), dict(size=key.size, md5=key.etag.strip('"')))

However Amazon specifies in the documentation that ETag might be or not MD5 hash:
"Entity tag that identifies the newly created object's data. Objects with different object data will have different entity tags. The entity tag is an opaque string. The entity tag may or may not be an MD5 digest of the object data. If the entity tag is not an MD5 digest of the object data, it will contain one or more nonhexadecimal characters and/or will consist of less than 32 or more than 32 hexadecimal digits."
http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadComplete.html

this is how boto checks if etag is md5 : if (re.match('^"[a-fA-F0-9]{32}"$', key.etag)):
so you might come up of generating md5 for those files differently

**Comment by nathancahill** [comment]
@justindowning Sorry for the long delay. This should be fixed in the latest build (1.0.5)
